### PR TITLE
Add tests for AppSettings

### PR DESCRIPTION
**Pull request by ivangrek**
## Proposed changes

- Less magic values
- More tests

## Test methodology 

- Manual
- Automation

## Test environment(s) 

- Git Extensions 3.4.3.9999
- Build d4b0f48bbf3e39a71f5d7ff5231be5678d4aa0f1
- Git 2.28.0.windows.1
- Microsoft Windows NT 10.0.19041.0
- .NET Framework 4.8.4220.0
- DPI 96dpi (no scaling)

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

**Comment by RussKie** [comment]
There is a lot of effort, but I can't sign off on these changes for a number of reasons.

We are adding a lot of code that has no user value, adds additional maintenance overhead, and is for test purposes only.
That is, to add a new setting we need to add a new const, and then reference it in the code.

We are also adding public const.  Const values get inlined during compilation times, which essentially binds us to these values forever - we won't be able to change those without risking breaking plugins, that may have taken dependencies on these values.
In the last 4 years alone we have tweaked/or talked about flipping defaults on selected settings. We won't be able to do this.

Perhaps what we could do instead is to update tests to assert the current default value without relying on typed "defaults", e.g. (pseudo code):
```cs
        private static IEnumerable<(string propertyName, object defaultTypeValue)> GetDefaultValue()
        {
            yield return (nameof(AppSettings.TelemetryEnabled), true);
            ....
        }
```

**Comment by RussKie** [comment]
Also something on my radar (but hasn't reached the top priority) - look at how we implemented test accessors in Windows Forms repo: https://github.com/dotnet/winforms/pull/3225. It employs reflection as oppose to how we implemented them here (via TestAccessor structs).
This may be specifically useful in your scenarios.

**Comment by ivangrek** [comment]
I tried to link defaults and settings. This is also part of the test.

**Comment by ivangrek** [comment]
> Perhaps what we could do instead is to update tests to assert the current default value without relying on typed "defaults", e.g. (pseudo code):

We will double the number of magic values ​​in the code. )

**Comment by RussKie** [comment]
> We will double the number of magic values ​​in the code. )

Yes, and no - values for the default values (a mouth full) are not valuable on their own. If values need changing we'll have a discussion and an associated change, that will include the change in AppSettings and a change in an associated test.

**Comment by ivangrek** [comment]
Removed constants

**Comment by ivangrek** [comment]
At least weirdness has already been identified.

> `yield return (nameof(AppSettings.TelemetryEnabled), true);`

And in fact.

`yield return (nameof(AppSettings.TelemetryEnabled), null);`

As I think it should be a non-nullable boolean.

File: UnitTests/GitCommands.Tests/Settings/AppSettingsTests.cs [hunk 1]
```diff
@@ -0,0 +1,641 @@
...
+                        var propertyType = Nullable.GetUnderlyingType(property.PropertyType) ?? property.PropertyType;
+
+                        if (valueType == propertyType)
+                        {
+                            yield return new object[] { propertyName, value, defaultValue, isSetting };
+                        }
+                    }
+                }
+            }
+        }
```
Comment (RussKie) [inline UnitTests/GitCommands.Tests/Settings/AppSettingsTests.cs:131]: 💭  I don't typically like using local functions, but I think here is a good scenario to use them and move all the utility methods into `TestCases` so they don't pollute the class namespace.

**Review by RussKie — CHANGES_REQUESTED** [review]
👍 with few comments

File: UnitTests/GitCommands.Tests/Settings/AppSettingsTests.cs [hunk 2]
```diff
@@ -0,0 +1,641 @@
...
+                {
+                    var property = typeof(AppSettings).GetProperty(propertyName);
+
+                    if (value is null)
+                    {
+                        if (defaultTypeValue is null)
+                        {
+                            yield return new object[] { propertyName, value, defaultValue, isSetting };
+                        }
+                    }
```
[the remark this answers is not in the document]
Comment (ivangrek) [reply UnitTests/GitCommands.Tests/Settings/AppSettingsTests.cs:118]: Null value can apply only for nullable type whre defaultTypeValue is null.

[the remark this answers is not in the document]
Comment (RussKie) [reply UnitTests/GitCommands.Tests/Settings/AppSettingsTests.cs:118]: So we should invert to 
```cs
if (defaultTypeValue is null && value is null) { ... }
else { ... }
```
?

**Review by RussKie — APPROVED** [review]

**Comment by RussKie** [comment]
🚀

**Pull request merged**